### ISSUE TITLE
cuda.memory_pool unit test failing with clang/7+cuda/9.2, Kepler35 arch

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#3589]
The `cuda.memory_pool` test began failing following merge of PR #3580

I reproduced the failure on kokkos-dev and did a bisection on merge commits to confirm,

Failing SHA: 1b432961fa2a5b7435d340425bb62988a98317cc

Passing SHA: 72375bfd47bb3101b4a698dc552e5583a57f393f

@crtrott are you able to take a look? Looks like more whack-a-mole after the zero-length reduction fixes :(

**Test failure output:**
```
18:45:01 /home/jenkins/slave/workspace/Kokkos_develop_kokkos-dev_clang_701_cuda_92/kokkos/core/unit_test/TestMemoryPool.hpp:326: Failure
18:45:01 Value of: long(stats.consumed_blocks)
18:45:01   Actual: 46666
18:45:01 Expected: (nfill - ndel)
18:45:01 Which is: 48829
18:45:01 [  FAILED  ] cuda.memory_pool (22 ms)
```

**Sample reproducer instructions (kokkos-dev):**
```bash
module load sems-env sems-cmake/3.12.2 sems-clang/7.0.1 sems-cuda/9.2

$KOKKOS_PATH/generate_makefile.bash --with-devices=Cuda,OpenMP --arch=Kepler35 --compiler=clang++ --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="14" --ldflags="" --with-cuda --kokkos-path=$KOKKOS_PATH  --no-examples

# Corresponding cmake line
cmake -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_CXX_FLAGS=”-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized  --gcc-toolchain=/projects/sems/install/rhel7-x86_64/sems/compiler/gcc/5.3.0/base --cuda-path=/projects/sems/install/rhel7-x86_64/sems/compiler/cuda/9.2/base” -DKokkos_ENABLE_OPENMP=ON -DKokkos_ENABLE_CUDA=ON -DKokkos_ARCH_KEPLER35=ON -DKokkos_ENABLE_TESTS=ON -DKokkos_ENABLE_EXAMPLES=OFF -DCMAKE_VERBOSE_MAKEFILE=ON -DCMAKE_CXX_EXTENSIONS=OFF -DCMAKE_CXX_STANDARD=14 -DCMAKE_BUILD_TYPE=RELEASE $KOKKOS_PATH
```

**Comment by ndellingwood** [comment]
We should probably add the `AAAAHHH GRRRR` label...

**Comment by janciesko** [comment]
Reproduced, taking a look.

**Comment by janciesko** [comment]
Draft PR https://github.com/kokkos/kokkos/pull/3631

**Comment by ndellingwood** [comment]
Merged during part of 3.2.1 release
